processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -384,55 +411,127 @@ public int add(InputRow row) throws IndexSizeExceededException
       throw new IAE("Cannot add row[%s] because it is below the minTimestamp[%s]", row, new DateTime(minTimestamp));
     }
 
+    String[][] dims = null;
     final List<String> rowDimensions = row.getDimensions();
 
-    String[][] dims;
-    List<String[]> overflow = null;
-    synchronized (dimensionOrder) {
-      dims = new String[dimensionOrder.size()][];
-      for (String dimension : rowDimensions) {
-        List<String> dimensionValues = row.getDimension(dimension);
-
-        // Set column capabilities as data is coming in
-        ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
-        if (capabilities == null) {
-          capabilities = new ColumnCapabilitiesImpl();
-          capabilities.setType(ValueType.STRING);
-          columnCapabilities.put(dimension, capabilities);
+    for (String dimension : rowDimensions) {
+      List<String> dimensionValues = row.getDimension(dimension);
+      final Boolean raceWinner;
+      if (dimensionOrder.containsKey(dimension)) {
+        // Common case, don't even try to lock
+        raceWinner = false;
+      }else{
+        final Lock dimLock = dimensionSynchronizer.get(dimension);
+        dimLock.lock();
+        try {
+          // If we didn't win the race, populate cache values in the same codepath below as if we never tried to lock
+          if(raceWinner = !dimensionOrder.containsKey(dimension)){
+            // we won a race, add a new dimension
+            // Set column capabilities as data is coming in if it wasn't set in constructor
+            ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
+            if (capabilities == null) {
+              capabilities = new ColumnCapabilitiesImpl();
+              capabilities.setType(ValueType.STRING);
+              ColumnCapabilitiesImpl other = columnCapabilities.putIfAbsent(dimension, capabilities);
+              if (other != null) {
+                throw new ISE("Bad columnCapabilites state");
+              }
+            }
+            if (dimensionValues.size() > 1) {
+              capabilities.setHasMultipleValues(true);
+            }
+
+            final Integer orderID = orderIncrementer.getAndIncrement();
```

Comment:
Because the objects are infinitely more useful and boxing/unboxing is almost never an issue outside of an array.